### PR TITLE
feat: optimise nft lucky drop

### PR DESCRIPTION
**Pull request by hanszhoou**
## Description
https://mask.atlassian.net/browse/MF-3245
https://mask.atlassian.net/browse/MF-3215
https://mask.atlassian.net/browse/MF-3256
https://mask.atlassian.net/browse/MF-3257

Closes # (NO_ISSUE)

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/web3-hooks/base/src/useNonFungibleAsset.ts [hunk 1]
```diff
@@ -18,6 +18,6 @@ export function useNonFungibleAsset<S extends 'all' | void = void, T extends Net
 
     return useAsyncRetry<Web3Helper.NonFungibleAssetScope<S, T> | undefined>(async () => {
         if (!hub) return
-        return hub.getNonFungibleAsset?.(address ?? '', id ?? '')
+        return hub.getNonFungibleAsset?.(address ?? '', id ?? '', options)
     }, [address, id, hub])
```
Comment (guanbinrui) [inline packages/web3-hooks/base/src/useNonFungibleAsset.ts]:
```suggestion
    }, [address, id, hub, JSON.stringify(options)])
```

File: packages/web3-hooks/evm/src/useNonFungibleOwnerTokens.ts [hunk 2]
```diff
@@ -19,8 +19,26 @@ export function useNonFungibleOwnerTokens(
     const nonFungibleTokenContract = useERC721TokenContract(chainId, contractAddress ?? '')
     const connection = useWeb3Connection(NetworkPluginID.PLUGIN_EVM, { chainId })
 
+    const { value: collectibles = EMPTY_LIST, done: loadFinish } = useNonFungibleAssets(
+        NetworkPluginID.PLUGIN_EVM,
+        SchemaType.ERC721,
+        { chainId, account: ownerAccount },
+    )
+
+    allListRef.current = collectibles
```
Comment (guanbinrui) [inline packages/web3-hooks/evm/src/useNonFungibleOwnerTokens.ts]: Don't do this in a hook directly.

File: packages/mask/src/plugins/RedPacket/SNSAdaptor/RedPacketERC721Form.tsx [hunk 3]
```diff
@@ -273,6 +273,8 @@ export function RedPacketERC721Form(props: RedPacketERC721FormProps) {
 
     const { RED_PACKET_NFT_ADDRESS } = useNftRedPacketConstants(chainId)
 
+    const { value: customCollections = [] } = useNFTRedpacketCustomCollections(account, chainId)
```
Comment (UncleBill) [inline packages/mask/src/plugins/RedPacket/SNSAdaptor/RedPacketERC721Form.tsx]: You are passing it down as a prop, set the fallback to stable EMPTY_LIST.

```suggestion
    const { value: customCollections = EMPTY_LIST } = useNFTRedpacketCustomCollections(account, chainId)
```

File: packages/mask/src/plugins/Wallet/SNSAdaptor/SelectNftContractDialog.tsx [hunk 4]
```diff
@@ -134,21 +130,25 @@ export function SelectNftContractDialog(props: SelectNftContractDialogProps) {
...
-                    name: x.name,
-                    symbol: x.symbol,
-                    baseURI: x.iconURL,
-                    iconURL: x.iconURL,
-                    balance: x.tokensTotal,
-                } as NonFungibleTokenContract<ChainId, SchemaType>),
-        )
+    const contractList = uniqBy(
+        assets
+            .concat(customCollections)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Wallet/SNSAdaptor/SelectNftContractDialog.tsx]: Move to `useNonFungibleCollections`.

File: packages/mask/src/plugins/RedPacket/SNSAdaptor/hooks/useNFTRedpacketCustomCollections.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import urlcat from 'urlcat'
+import { mapKeys } from 'lodash-es'
+import { useAsyncRetry } from 'react-use'
+import { fetchJSON } from '@masknet/web3-providers/helpers'
+import type { ChainId, SchemaType } from '@masknet/web3-shared-evm'
+import type { NonFungibleCollection } from '@masknet/web3-shared-base'
+
+const DSEARCH_BASE_URL = 'https://dsearch.mask.r2d2.to'
+
+export function useNFTRedpacketCustomCollections(owner: string, chainId: ChainId) {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/RedPacket/SNSAdaptor/hooks/useNFTRedpacketCustomCollections.ts]: Merge into `useNonFungibleCollections`.

File: packages/web3-hooks/base/src/useNonFungibleCollections.ts [hunk 6]
```diff
@@ -25,6 +30,22 @@ export function useNonFungibleCollections<S extends 'all' | void = void, T exten
                 ...options,
             })
         })
-        return asyncIteratorToArray(iterator)
-    }, [account, hub, JSON.stringify(options)])
+
+        return attemptUntil(
+            [
+                async () => asyncIteratorToArray(iterator),
+                async () => {
```
Comment (guanbinrui) [inline packages/web3-hooks/base/src/useNonFungibleCollections.ts:37]: Move to the Hub of EVM. This hook is shared across all networks.

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**